In [1]:

import sys

# Insert at 1, 0 is the script path (or '' in REPL)
sys.path.insert(1, '../py')

In [2]:

%run ../../load_magic/environment.py
%run ../../load_magic/paths.py
%run ../../load_magic/lists.py
import spiral_utils
u = spiral_utils.StraussHoweUtilities()
pd = spiral_utils.pd
s = u.s
%pprint
notebook_path = get_notebook_path()
print(notebook_path)
print(['s.{}'.format(fn) for fn in dir(s) if not fn.startswith('_')])
print(['u.{}'.format(fn) for fn in dir(u) if not fn.startswith('_')])
dir()

Pretty printing has been turned OFF
C:\Users\dev\Documents\repositories\notebooks\Strauss–Howe\ipynb\Calvinism.ipynb
['s.attempt_to_pickle', 's.data_csv_folder', 's.data_folder', 's.encoding_type', 's.load_csv', 's.load_dataframes', 's.load_object', 's.save_dataframes', 's.saves_csv_folder', 's.saves_folder', 's.saves_pickle_folder', 's.store_objects']
['u.add_fashion_image', 'u.add_fashion_saeculum_image', 'u.add_guide', 'u.add_saeculum_image', 'u.add_spiral_labels', 'u.archetypes_df', 'u.archimedes_spiral', 'u.babbitt_min_year', 'u.babbitt_random_year', 'u.babbitt_years_tuple', 'u.bare_movie_folder', 'u.bernoulli_spiral', 'u.black_tuple', 'u.colors_dict_to_df', 'u.conjunctify_list', 'u.convert_age_to_y', 'u.convert_years_to_x', 'u.create_xy_list', 'u.cycles_image', 'u.diagonal_movie_folder', 'u.display_test_colors', 'u.distance_between', 'u.distance_from_black', 'u.distance_from_blue', 'u.distance_from_cyan', 'u.distance_from_green', 'u.distance_from_kbcg_face', 'u.distance_from_krmb

['Config', 'In', 'Out', 'RandomForestClassifier', 'SequenceMatcher', '_', '__', '___', '__builtin__', '__builtins__', '__doc__', '__loader__', '__name__', '__nonzero__', '__package__', '__spec__', '_dh', '_i', '_i1', '_i2', '_ih', '_ii', '_iii', '_oh', 'check_4_doubles', 'check_for_typos', 'conjunctify_list', 'copyfile', 'encoding', 'exit', 'get_classifier', 'get_data_structs_df', 'get_datastructure_prediction', 'get_dir_tree', 'get_git_lfs_track_commands', 'get_importances', 'get_input_sample', 'get_ipython', 'get_module_version', 'get_notebook_path', 'get_specific_gitignore_files', 'get_struct_name', 'humanize_bytes', 'ipykernel', 'json', 'jupyter_config_dir', 'notebook_path', 'notebookapp', 'os', 'pd', 'preprocess_data', 'print_all_files_ending_starting_with', 'print_all_files_ending_with', 'print_all_files_starting_with', 'quit', 're', 'remove_empty_folders', 's', 'similar', 'spiral_utils', 'subprocess', 'sys', 'time', 'u', 'urllib']

In [31]:

# Create a comprehensive dataset to train on
import wikipedia

def f():

    return None

class ObjectClass(object):
    def __init__(self):
        self.its = ''
    def im_a_method(self):
        self.its = 'method'

c = ObjectClass()
data_struct_list = [list, tuple, set, dict, str, pd.DataFrame, pd.np.ndarray,
                    bytearray, range, os, pd, f, ObjectClass, c, c.im_a_method,
                    True, None, wikipedia.BeautifulSoup, wikipedia.Decimal,
                    wikipedia.util]
data_structs_df = get_data_structs_df(data_struct_list=data_struct_list)
X, y = preprocess_data(data_structs_df)
#print('The shape of X is:', X.shape)
#print('The shape of y is:', y.shape)
clf = get_classifier(X, y)
get_importances(data_structs_df, clf)[:10]

f_str = '{} is a type of {}, named a {}, but predicted to be a {}'
for data_struct in data_struct_list:
    print(f_str.format(str(data_struct), type(data_struct), get_struct_name(data_struct),
                       get_datastructure_prediction(df=data_structs_df, clf=clf, func=data_struct)))
print(data_structs_df.index, y)

<class 'list'> is a type of <class 'type'>, named a list, but predicted to be a list
<class 'tuple'> is a type of <class 'type'>, named a tuple, but predicted to be a tuple
<class 'set'> is a type of <class 'type'>, named a set, but predicted to be a set
<class 'dict'> is a type of <class 'type'>, named a dictionary, but predicted to be a dictionary
<class 'str'> is a type of <class 'type'>, named a string, but predicted to be a string
<class 'pandas.core.frame.DataFrame'> is a type of <class 'type'>, named a import, but predicted to be a boolean
<class 'numpy.ndarray'> is a type of <class 'type'>, named a import, but predicted to be a boolean
<class 'bytearray'> is a type of <class 'type'>, named a bytearray, but predicted to be a bytearray
<class 'range'> is a type of <class 'type'>, named a range, but predicted to be a range
<module 'os' from 'C:\\Users\\dev\\Anaconda3\\lib\\os.py'> is a type of <class 'module'>, named a module, but predicted to be a module
<module 'pandas' from 'C:

In [32]:

f_str = '{} ({}) is a {}'
for (s_str, data_struct) in [('wikipedia.{}'.format(fn), eval('wikipedia.{}'.format(fn))) for fn in dir(wikipedia) if not fn.startswith('_')]:
    prediction = get_datastructure_prediction(df=data_structs_df, clf=clf, func=data_struct)
    if '__code__' in dir(data_struct):
        s_str = '{}{}'.format(s_str, data_struct.__code__.co_varnames)
        s_str = ''.join(s_str.split("'"))
    print(f_str.format(s_str, get_struct_name(data_struct), prediction))

wikipedia.API_URL (string) is a string
wikipedia.BeautifulSoup (import) is a dictionary
wikipedia.Decimal (import) is a import
wikipedia.DisambiguationError (import) is a instantiation
wikipedia.HTTPTimeoutError (import) is a instantiation
wikipedia.ODD_ERROR_MESSAGE (string) is a string
wikipedia.PageError (import) is a instantiation
wikipedia.RATE_LIMIT (boolean) is a boolean
wikipedia.RATE_LIMIT_LAST_CALL (NoneType) is a NoneType
wikipedia.RATE_LIMIT_MIN_WAIT (NoneType) is a NoneType
wikipedia.RedirectError (import) is a instantiation
wikipedia.USER_AGENT (string) is a string
wikipedia.WikipediaException (import) is a instantiation
wikipedia.WikipediaPage (import) is a instantiation
wikipedia.cache (import) is a instantiation
wikipedia.datetime (import) is a NoneType
wikipedia.debug(fn, wrapper) (function) is a function
wikipedia.donate(webbrowser,) (function) is a function
wikipedia.exceptions (module) is a module
wikipedia.geosearch (wikipedia.util.cache) is a instantiation
wikipe

In [42]:

%run ../../load_magic/soup.py

print(['wikipedia.requests.urllib3.{}'.format(fn) for fn in dir(wikipedia.requests.urllib3) if not fn.startswith('_')])
print(['wikipedia.requests.{}'.format(fn) for fn in dir(wikipedia.requests) if not fn.startswith('_')])
dir()

['wikipedia.requests.urllib3.HTTPConnectionPool', 'wikipedia.requests.urllib3.HTTPResponse', 'wikipedia.requests.urllib3.HTTPSConnectionPool', 'wikipedia.requests.urllib3.PoolManager', 'wikipedia.requests.urllib3.ProxyManager', 'wikipedia.requests.urllib3.Retry', 'wikipedia.requests.urllib3.Timeout', 'wikipedia.requests.urllib3.absolute_import', 'wikipedia.requests.urllib3.add_stderr_logger', 'wikipedia.requests.urllib3.connection', 'wikipedia.requests.urllib3.connection_from_url', 'wikipedia.requests.urllib3.connectionpool', 'wikipedia.requests.urllib3.contrib', 'wikipedia.requests.urllib3.disable_warnings', 'wikipedia.requests.urllib3.encode_multipart_formdata', 'wikipedia.requests.urllib3.exceptions', 'wikipedia.requests.urllib3.fields', 'wikipedia.requests.urllib3.filepost', 'wikipedia.requests.urllib3.get_host', 'wikipedia.requests.urllib3.logging', 'wikipedia.requests.urllib3.make_headers', 'wikipedia.requests.urllib3.packages', 'wikipedia.requests.urllib3.poolmanager', 'wikipedi

['Config', 'In', 'ObjectClass', 'Out', 'Path', 'RandomForestClassifier', 'SequenceMatcher', 'X', '_', '_11', '_13', '_19', '_2', '_20', '_21', '_22', '_23', '_28', '_29', '_30', '_34', '_35', '_36', '_38', '_41', '_5', '_6', '_7', '_9', '__', '___', '__builtin__', '__builtins__', '__doc__', '__loader__', '__name__', '__nonzero__', '__package__', '__spec__', '_dh', '_i', '_i1', '_i10', '_i11', '_i12', '_i13', '_i14', '_i15', '_i16', '_i17', '_i18', '_i19', '_i2', '_i20', '_i21', '_i22', '_i23', '_i24', '_i25', '_i26', '_i27', '_i28', '_i29', '_i3', '_i30', '_i31', '_i32', '_i33', '_i34', '_i35', '_i36', '_i37', '_i38', '_i39', '_i4', '_i40', '_i41', '_i42', '_i5', '_i6', '_i7', '_i8', '_i9', '_ih', '_ii', '_iii', '_oh', 'bs', 'c', 'check_4_doubles', 'check_for_typos', 'clf', 'conjunctify_list', 'copyfile', 'data_struct', 'data_struct_list', 'data_structs_df', 'display_regex', 'encoding', 'exit', 'f', 'f_str', 'get_classifier', 'get_data_structs_df', 'get_datastructure_prediction', 'get_

In [38]:

wikipedia.search(query='Babbitt', results=10, suggestion=True)

(['Babbitt', 'Babbitt (novel)', 'Bruce Babbitt', 'Art Babbitt', 'Babbitt (alloy)', 'Milton Babbitt', 'Rain Man', 'Babbitt (surname)', 'Natalie Babbitt', 'Luke Babbitt'], None)

In [36]:

wikipedia.summary(title='Johannes Oecolampadius', sentences=3)

'Johannes Oecolampadius (also Œcolampadius, in German also Oekolampadius, Oekolampad; 1482 – 24 November 1531) was a German Protestant reformer in the Reformed tradition from the Electoral Palatinate. He was the leader of the Protestant faction in the Baden Disputation of 1526, and he was one of the founders of Protestant theology, engaging in disputes with Erasmus, Zwingli, Luther and Martin Bucer. Calvin adopted his view on the eucharist dispute  (against Luther).'

In [45]:

links_url = 'https://en.wikipedia.org/w/index.php?title=Special:WhatLinksHere/{}&limit=5000'.format('Calvinism')
page_soup = get_page_soup(links_url)

In [57]:

def get_bio_row_dict(page_soup):
    row_dict = {}
    for param_th in page_soup.find_all('th', scope='row'):
        if (param_th.text.strip()=='Baptised') or (param_th.text.strip()=='Born') or (param_th.text.strip()=='Died'):
            row_dict['page_title'] = page_title
        if (param_th.text.strip()=='Baptised') or (param_th.text.strip()=='Born'):
            born_td = param_th.find_next_sibling('td')
            born_list = born_regex.findall(born_td.text)
            if len(born_list) > 0:
                born_x = max(born_list, key=lambda x: len(x[0]))[0]
                born_x = int(bc_regex.sub(r'-\1', born_x))
                row_dict['birth_year'] = born_x
        if (param_th.text.strip()=='Died'):
            died_td = param_th.find_next_sibling('td')
            died_list = born_regex.findall(died_td.text)
            if len(died_list) > 0:
                died_x = max(died_list, key=lambda x: len(x[0]))[0]
                died_x = int(bc_regex.sub(r'-\1', died_x))
                row_dict['death_year'] = died_x
    return row_dict

In [62]:

ul_str = '<ul id="mw-whatlinkshere-list">'
ul_soup = page_soup.find(name='ul', id='mw-whatlinkshere-list')
li_list = ul_soup.find_all(name='li')

In [64]:

from random import shuffle

y_supply = list(range(100))
shuffle(y_supply)
born_regex = re.compile(r'(\d{3,4}( BC)?)')
bc_regex = re.compile(r'(\d+) BC')
pos_dict = {}
rows_list = []

def y():
    
    return (y_supply.pop())

In [76]:

calvinists_df = s.load_object('calvinists_df')
calvinists_df.columns

Index(['page_title', 'birth_year', 'death_year', 'birth_diff',
       'generations_archetype', 'turnings_archetype', 'generation_name',
       'saeculum_name'],
      dtype='object')

In [68]:

def create_poi_df(rows_list):
    df = pd.DataFrame(rows_list)
    match_series = df.duplicated(subset='page_title', keep='first')
    df = df[~match_series].sort_values('birth_year').dropna(axis='index',
                                                            how='all').reset_index(drop=True)
    df['birth_diff'] = df.birth_year.diff()
    
    return df

In [69]:

def enhance_poi_df(poi_df):
    generations_df = s.load_object('generations_df')
    f_str = '{}, {} {} {} of the {} generation in the {} saeculum.'
    for row_index, row_series in poi_df.iterrows():
        birth_year = row_series['birth_year']
        match_series = (generations_df.birth_year_begin <= birth_year)
        match_series = match_series & (generations_df.birth_year_end >= birth_year)
        df = generations_df[match_series]
        if df.shape[0] > 0:
            page_title = row_series.page_title
            generations_archetype = df.generations_archetype.tolist()[0]
            poi_df.loc[row_index, 'generations_archetype'] = generations_archetype
            turnings_archetype = df.turnings_archetype.tolist()[0]
            poi_df.loc[row_index, 'turnings_archetype'] = turnings_archetype
            generation_name = df.index.tolist()[0]
            poi_df.loc[row_index, 'generation_name'] = generation_name
            saeculum_name = df.saeculum_name.tolist()[0]
            poi_df.loc[row_index, 'saeculum_name'] = saeculum_name
            if generations_archetype[0] in ['A', 'I']:
                a_an = 'an'
            else:
                a_an = 'a'
            print(f_str.format(page_title, a_an, generations_archetype,
                               turnings_archetype, generation_name,
                               saeculum_name))

    return poi_df.sort_values('birth_year')

In [87]:

def get_wikipedia_page_soup(page_title):
    page_soup = None
    try:
        page_obj = wikipedia.WikipediaPage(title=page_title)
        page_html = page_obj.html()
        page_soup = bs(page_html, 'html.parser')
    except Exception as e:
        print(e)
    
    return page_soup

In [ ]:

#calvinists_df['linked_to_calvinism'] = False
for li_soup in li_list:
    link_tag = list(li_soup.children)[0]
    page_title = link_tag.attrs['title']
    page_soup = get_wikipedia_page_soup(page_title)
    if page_soup is not None:
        row_dict = get_bio_row_dict(page_soup)
        if len(row_dict) > 2:
            match_series = (calvinists_df.page_title == row_dict['page_title'])
            df = calvinists_df[match_series]
            if (df.shape[0] > 0) and not df.linked_to_calvinism.all():
                calvinists_df.loc[match_series, 'linked_to_calvinism'] = True
            else:
                df = create_poi_df([row_dict])
                df = enhance_poi_df(poi_df=df)
                df['linked_to_calvinism'] = True
                calvinists_df = calvinists_df.append(df).reset_index(drop=True)
s.store_objects(calvinists_df=calvinists_df)

In [92]:

def f(x):
    
    return str(x).startswith('User:')

match_series = calvinists_df.page_title.map(f)
calvinists_df = calvinists_df[~match_series].sort_values(['birth_year',
                                                          'death_year']).dropna(axis='index',
                                                                                how='all').reset_index(drop=True)
calvinists_df['birth_diff'] = calvinists_df.birth_year.diff()

In [95]:

match_series = calvinists_df.duplicated(subset=['page_title', 'birth_year', 'death_year'], keep='first')
calvinists_df = calvinists_df[~match_series].sort_values(['birth_year',
                                                          'death_year']).dropna(axis='index',
                                                                                how='all').reset_index(drop=True)
calvinists_df['birth_diff'] = calvinists_df.birth_year.diff()
calvinists_df

birth_diff  birth_year  death_year generation_name generations_archetype  \
0           NaN       100.0       165.0             NaN                   NaN   
1          30.0       130.0       202.0             NaN                   NaN   
2         127.0       257.0       331.0             NaN                   NaN   
3           3.0       260.0       339.0             NaN                   NaN   
4          12.0       272.0       337.0             NaN                   NaN   
..          ...         ...         ...             ...                   ...   
697         0.0      1948.0      2008.0       Baby Boom              Idealist   
698         2.0      1950.0         NaN       Baby Boom              Idealist   
699         1.0      1951.0      1997.0       Baby Boom              Idealist   
700         7.0      1958.0      2017.0       Baby Boom              Idealist   
701         6.0      1964.0         NaN            13th              Reactive   

     linked_to_calvinism                   page_title saeculum_name  \
0                   True                Justin Martyr           NaN   
1                   True                     Irenaeus           NaN   
2                   True      Gregory the Illuminator           NaN   
3                   True                     Eusebius           NaN   
4                   True        Constantine the Great           NaN   
..                   ...                          ...           ...   
697                 True              William Placher    Millennial   
698                False            Timothy J. Keller    Millennial   
699                 True                David Chilton    Millennial   
700                 True                    Mike Ovey    Millennial   
701                False  Michael Horton (theologian)    Millennial   

    turnings_archetype  
0                  NaN  
1                  NaN  
2                  NaN  
3                  NaN  
4                  NaN  
..                 ...  
697            Prophet  
698            Prophet  
699            Prophet  
700            Prophet  
701              Nomad  

[702 rows x 9 columns]

In [96]:

match_series = (calvinists_df.birth_diff == 0.0)
calvinists_df[match_series]

birth_diff  birth_year  death_year generation_name generations_archetype  \
26          0.0      1489.0      1562.0     Reformation              Idealist   
27          0.0      1489.0      1565.0     Reformation              Idealist   
33          0.0      1498.0      1552.0     Reformation              Idealist   
35          0.0      1499.0      1565.0     Reformation              Idealist   
38          0.0      1503.0      1559.0     Reformation              Idealist   
..          ...         ...         ...             ...                   ...   
685         0.0      1932.0      2019.0          Silent              Adaptive   
689         0.0      1939.0         NaN          Silent              Adaptive   
690         0.0      1939.0         NaN          Silent              Adaptive   
695         0.0      1943.0      2016.0       Baby Boom              Idealist   
697         0.0      1948.0      2008.0       Baby Boom              Idealist   

     linked_to_calvinism           page_title saeculum_name turnings_archetype  
26                  True  François de Tournon   Reformation            Prophet  
27                  True        William Farel   Reformation            Prophet  
33                  True     Andreas Osiander   Reformation            Prophet  
35                  True         Pope Pius IV   Reformation            Prophet  
38                  True      Robert Estienne   Reformation            Prophet  
..                   ...                  ...           ...                ...  
685                 True       Norman Geisler   Great Power             Artist  
689                False    John F. MacArthur   Great Power             Artist  
690                False       David F. Wells   Great Power             Artist  
695                 True      István Komáromi    Millennial            Prophet  
697                 True      William Placher    Millennial            Prophet  

[335 rows x 9 columns]

In [97]:

s.store_objects(calvinists_df=calvinists_df)

Pickling to C:\Users\dev\Documents\repositories\notebooks\Strauss–Howe\saves\pickle\calvinists_df.pickle


In [ ]:

!start %windir%\explorer.exe "{os.path.abspath(os.path.join(notebook_path, '../'))}"

In [ ]:

def get_toc_div(page_title):
    page_obj = wikipedia.WikipediaPage(title=page_title)
    page_html = page_obj.html()
    page_soup = bs(page_html, 'html.parser')
    toc_div = page_soup.find_all('div', id='toc')[0]
    
    return toc_div

In [61]:

def get_poi_rows_list(toc_div):
    rows_list = []
    for sib_tag in toc_div.find_previous_siblings():
        #print(['sib_tag.{}'.format(fn) for fn in dir(sib_tag) if not fn.startswith('_')])
        if len(sib_tag.attrs) == 0:
            for a_tag in sib_tag.find_all(name='a', title=True):
                page_title = a_tag.attrs['title']
                page_soup = get_wikipedia_page_soup(page_title)
                row_dict = get_bio_row_dict(page_soup)
                rows_list.append(row_dict.copy())
    
    return rows_list

In [ ]:

def get_generations_archetype(row_series):
    generations_archetype = row_series.generations_archetype
    if generations_archetype[0] in ['A', 'I']:
        a_an = 'an'
    else:
        a_an = 'a'
    
    return a_an, generations_archetype

In [ ]:

toc_div = get_toc_div(page_title='Calvinism')
rows_list = get_poi_rows_list(toc_div)
calvinists_df = create_poi_df(rows_list)
calvinists_df = enhance_poi_df(poi_df=calvinists_df)
s.store_objects(calvinists_df=calvinists_df)

In [ ]:

toc_div = get_toc_div(page_title='Presbyterianism')
rows_list = get_poi_rows_list(toc_div)
presbyterians_df = create_poi_df(rows_list)
presbyterians_df = enhance_poi_df(poi_df=presbyterians_df)
s.store_objects(presbyterians_df=presbyterians_df)

In [ ]:

style_dict = {'font-style': 'normal', 'font-variant': 'normal', 'font-weight': 'normal', 'font-stretch': 'normal',
              'line-height': '0%', 'font-family': 'Arial', '-inkscape-font-specification': 'Arial, Normal',
              'text-align': 'start', 'writing-mode': 'lr-tb', 'text-anchor': 'start', 'fill': '#000000', 'fill-opacity': '1'}
def get_style_list(style_dict):
    style_list = []
    for key, value in style_dict.items():
        if ' ' in value:
            style_list.append("{}:'{}'".format(key, value))
        else:
            style_list.append('{}:{}'.format(key, value))
    
    return style_list

In [ ]:

from scipy.optimize import curve_fit

def func(x, a, b):
    return a * x + b

year_data = pd.np.linspace(1500, 2000, 6)
#print(['pd.np.{}'.format(fn) for fn in dir(pd.np) if 'arr' in fn.lower()])
px_data = pd.np.array(object=[23.000, 173.599, 323.518, 473.479, 623.503, 773.573])
popt, pcov = curve_fit(func, year_data, px_data)
line_2d_obj = plt.plot(year_data, px_data, 'b-', label='data')

def get_horiz_px(year):
    
    return popt[0]*year + popt[1]

line_2d_obj = plt.plot(year_data, get_horiz_px(year_data), 'r-',
                       label='fit: horiz_pos = %5.3f * birth_year + %5.3f' % tuple(popt))
plt.xlabel('Birth Year')
plt.ylabel('Horizontal Position')
legend_obj = plt.legend()

In [ ]:

style_dict = {'opacity': '0.92000002', 'fill': '#cfaeff', 'fill-opacity': '0.43786986', 'fill-rule': 'evenodd',
              'stroke': 'none', 'stroke-width': '0.55682105', 'stroke-miterlimit': '4', 'stroke-dasharray': 'none',
              'stroke-dashoffset': '0', 'stroke-opacity': '1'}
style_list = get_style_list(style_dict.copy())
rect_str = '''
        <rect
           style="{}"
           id="rect{{}}"
           width="{{}}"
           height="720"
           x="{{}}"
           y="-215.64764"
           ry="0"
           inkscape:label="{{}} Rectangle" />'''.format(';'.join(style_list))

In [ ]:

saecula_df.loc['Great Power', 'crisis_begin_year'] = 1929
saecula_df.loc['Great Power', 'crisis_end_year'] = 1946
s.store_objects(saecula_df=saecula_df)

In [ ]:

flowroot_style_dict = {'font-style': 'normal', 'font-variant': 'normal', 'font-weight': 'normal', 'font-stretch': 'normal',
                       'font-size': '8px', 'line-height': '1.25', 'font-family': 'Arial',
                       '-inkscape-font-specification': "Arial, Normal", 'font-variant-ligatures': 'normal',
                       'font-variant-caps': 'normal', 'font-variant-numeric': 'normal', 'font-feature-settings': 'normal',
                       'text-align': 'center', 'letter-spacing': '0px', 'word-spacing': '0px', 'writing-mode': 'lr-tb',
                       'text-anchor': 'middle', 'fill': '#000000', 'fill-opacity': '1', 'stroke': 'none'}
flowroot_style_list = get_style_list(flowroot_style_dict.copy())
flowregion_style_dict = {'font-style': 'normal', 'font-variant': 'normal', 'font-weight': 'normal', 'font-stretch': 'normal',
                         'font-size': '8px', 'font-family': 'Arial',
                         '-inkscape-font-specification': "Arial, Normal", 'font-variant-ligatures': 'normal',
                         'font-variant-caps': 'normal', 'font-variant-numeric': 'normal', 'font-feature-settings': 'normal',
                         'text-align': 'center', 'writing-mode': 'lr-tb', 'text-anchor': 'middle'}
flowregion_style_list = get_style_list(flowregion_style_dict.copy())
rect_style_dict = {'font-style': 'normal', 'font-variant': 'normal', 'font-weight': 'normal', 'font-stretch': 'normal',
                   'font-size': '8px', 'font-family': 'Arial', '-inkscape-font-specification': "Arial, Normal",
                   'font-variant-ligatures': 'normal', 'font-variant-caps': 'normal',
                   'font-variant-numeric': 'normal', 'font-feature-settings': 'normal', 'text-align': 'center',
                   'writing-mode': 'lr-tb', 'text-anchor': 'middle'}
rect_style_list = get_style_list(rect_style_dict.copy())
flowpara_str = '''<flowPara
             id="flowPara{}">{}</flowPara>'''
header_str = '''
        <flowRoot
           inkscape:label="{{}} Text"
           transform="translate(0,0)"
           style="{}"
           id="flowRoot{{}}"
           xml:space="preserve"><flowRegion
             style="{}"
             id="flowRegion{{}}"><rect
               style="{}"
               x="{{}}"
               y="-245.31369"
               height="{{}}"
               width="65"
               id="rect{{}}" /></flowRegion>{{}}</flowRoot>'''.format(';'.join(flowroot_style_list),
                                                                      ';'.join(flowregion_style_list),
                                                                      ';'.join(rect_style_list))

In [ ]:

print(header_str)

In [ ]:

match_series = ~saecula_df.awakening_begin_year.isnull() & ~saecula_df.awakening_end_year.isnull()
flow_regex = re.compile(r'<g(\s+)inkscape:label="Awakenings Text"([^>]+)>(.+)</flowRoot>(\s+)</g>', re.DOTALL)
calvinists_file_path = os.path.join(s.saves_folder, 'svg', 'Calvinists.svg')
copy_file_path = os.path.join(s.saves_folder, 'svg', 'Calvinists - Copy.svg')
if os.path.exists(copy_file_path):
    if os.path.exists(calvinists_file_path):
        os.remove(calvinists_file_path)
    copyfile(copy_file_path, calvinists_file_path)
with open(calvinists_file_path, 'r') as f:
    xml_str = f.read()
header_xml_list = []
for row_index, row_series in saecula_df[match_series].iterrows():
    awakening_name = row_series.awakening_name
    awakening_begin_year = int(row_series.awakening_begin_year)
    awakening_end_year = int(row_series.awakening_end_year)
    begin_pos = get_horiz_px(awakening_begin_year)
    end_pos = get_horiz_px(awakening_end_year)
    width = end_pos - begin_pos
    id = '-' + '-'.join(row_index.lower().split(' '))
    if str(awakening_name) == 'nan':
        awakening_name = '{} Awakening'.format(row_index)
    #print(rect_str.format(id, width, begin_pos, awakening_name))
    #print('{} ({}-{})'.format(awakening_name, awakening_begin_year, awakening_end_year))
    awakening_name_list = awakening_name.split(' ')
    awakening_flowpara_str = ''.join([flowpara_str.format('{}-{}'.format(id, w.lower()),
                                                          w) for w in awakening_name_list])
    header_xml = header_str.format(awakening_name, id, id, begin_pos-15, len(awakening_name_list)*11,
                                   id, awakening_flowpara_str)
    header_xml_list.append(header_xml)
header_text_str = wle + wle.join(header_xml_list)
regex_sub_str = r'<g\g<1>inkscape:label="Awakenings Text"\g<2>>{}\g<4></g>'.format(header_text_str)
if flow_regex.search(xml_str):
    xml_str = flow_regex.sub(regex_sub_str, xml_str, 1)
    with open(calvinists_file_path, 'w') as f:
        print(xml_str.strip(), file=f)

In [ ]:

f_str = '{} from {} to {} ({} total)'
d_str = '{}, {} {} {} of the {} generation in the {} saeculum.'

In [ ]:

p_str = '''    <path
       style="stroke:#000000;stroke-width:2;stroke-miterlimit:4;stroke-dasharray:none"
       d="m {:0.6f},{:0.1f} c {:0.6f},0 {:0.6f},0 {:0.6f},0 h -6.20302"
       id="path{}"
       inkscape:connector-curvature="0"
       inkscape:label="{}">
      <desc
         id="desc{}">{}</desc>
      <title
         id="title{}">{}</title>
    </path>'''
paths_file_path = os.path.join(s.saves_folder, 'xml', 'Calvinist Paths.xml')

In [ ]:

style_list = get_style_list(style_dict.copy())
t_str = '''    <text
       x="{{}}"
       y="{{}}"
       id="text{{}}"
       xml:space="preserve"
       style="{}"
       inkscape:label="{{}}"><tspan
         sodipodi:role="line"
         id="tspan{{}}"
         x="{{}}"
         y="{{}}"
         style="font-size:13px;line-height:1.25">{{}}</tspan></text>'''.format(';'.join(style_list))
text_file_path = os.path.join(s.saves_folder, 'xml', 'Calvinist Text.xml')

In [ ]:

def get_birth_label_pos(row_series):
    birth_label = int(row_series.birth_year)
    if birth_label < 1500:
        birth_year = 1500
    else:
        birth_year = birth_label
    birth_pos = get_horiz_px(birth_year)
    
    return birth_label, birth_pos

In [ ]:

def get_death_label_pos(row_series):
    death_label = row_series.death_year
    if str(death_label) == 'nan':
        death_year = 2020
        death_label = 'present'
    else:
        death_label = int(death_label)
        death_year = death_label
    death_pos = get_horiz_px(death_year)
    
    return death_label, death_pos

In [ ]:

match_series = (calvinists_df.death_year >= 1500) | calvinists_df.death_year.isnull()
with open(paths_file_path, 'w') as f:
    print('', file=f)
with open(text_file_path, 'w') as f:
    print('', file=f)
paths_xml_list = []
texts_xml_list = []
for row_index, row_series in calvinists_df[match_series].iterrows():
    birth_label, birth_pos = get_birth_label_pos(row_series)
    death_label, death_pos = get_death_label_pos(row_series)
    page_title = row_series.page_title
    diff_pos = death_pos - birth_pos
    turnings_archetype = row_series.turnings_archetype
    generation_name = row_series.generation_name
    saeculum_name = row_series.saeculum_name
    description = d_str.format(page_title, *get_generations_archetype(row_series),
                               turnings_archetype, generation_name,
                               saeculum_name)
    label = '{} ({}-{})'.format(page_title, birth_label, death_label)
    print(row_index, '{}'.format(label))
    id = row_index+10000
    x = birth_pos
    y = 523+row_index*30
    path_str = p_str.format(x, y, diff_pos, diff_pos, diff_pos, id,
                            label, id, description, id, label)
    paths_xml_list.append(path_str)
    with open(paths_file_path, 'a') as f:
        print(path_str, file=f)
    text_str = t_str.format(x, y-5, id, label, id, x, y, label)
    texts_xml_list.append(text_str)
    with open(text_file_path, 'a') as f:
        print(text_str, file=f)

In [ ]:

from shutil import copyfile

r_str = r'\s+</g>\s+<g\s+id="Text"\s+style="font-size:13px;line-height:125%;font-family:Arial;fill:#000000">'
g_regex = re.compile(r_str, re.MULTILINE)
branches_file_path = os.path.join(s.saves_folder, 'svg', 'Protestant_branches.svg')
copy_file_path = os.path.join(s.saves_folder, 'svg', 'Protestant_branches - Copy.svg')
if os.path.exists(copy_file_path):
    if os.path.exists(branches_file_path):
        os.remove(branches_file_path)
    copyfile(copy_file_path, branches_file_path)
with open(branches_file_path, 'r') as f:
    xml_str = f.read()
inter_group_str = '''
  </g>
  <g
     id="Text"
     style="font-size:13px;line-height:125%;font-family:Arial;fill:#000000">'''
wle = r'\r\n'
paths_text_str = wle + wle.join(paths_xml_list) + inter_group_str + wle + wle.join(texts_xml_list)
if g_regex.search(xml_str):
    xml_str = g_regex.sub(paths_text_str, xml_str, 1)
    with open(branches_file_path, 'w') as f:
        print(xml_str.strip(), file=f)

In [ ]:

print(p_str)

In [ ]:

saecula_df = s.load_object('saecula_df')
columns_list = ['awakening_name', 'awakening_climax_year']
saecula_df[columns_list]

In [ ]:

    <path
       style="fill:none;stroke:#000000;stroke-width:2px;stroke-linecap:butt;stroke-linejoin:miter;stroke-opacity:1"
       d="m 23.262092,473.23916 c 46.522647,0 46.522647,0 46.522647,0 h -6.203018"
       id="path46"
       inkscape:connector-curvature="0"
       inkscape:label="Johannes Oecolampadius (1500-1531)"
    ><title id="title46">Johannes Oecolampadius (1500-1531)</title></path>

# import seaborn as sns

# First order (linear) scatterplot
kwargs = dict(textcoords='offset points', ha='left', va='bottom',
              bbox=dict(boxstyle='round,pad=0.5', fc='yellow', alpha=0.5),
              arrowprops=dict(arrowstyle='->', connectionstyle='arc3,rad=0'))
fig1_fig = plt.figure(figsize=(12, 8))
merge_axes_subplot = sns.regplot(x='birth_year', y='born_y',
                                 scatter=True, data=calvinists_df)
    
# Add annotations
for label, x, y in zip(calvinists_df.page_title, calvinists_df.birth_year, calvinists_df.born_y):
    if x < 400:
        annotation = plt.annotate('{}'.format(label), xy=(x, y), xytext=(50, -10), **kwargs)
    elif x < 1600:
        annotation = plt.annotate('{}'.format(label), xy=(x, y), xytext=(-150, -10), **kwargs)
    else:
        annotation = plt.annotate('{}'.format(label), xy=(x, y), xytext=(-130, -10), **kwargs)

In [ ]:

def get_influences_list(page_title='Judith Butler'):
    influences_list = []
    try:
        page_obj = wikipedia.WikipediaPage(title=page_title)
        page_html = page_obj.html()
        page_soup = bs(page_html, 'html.parser')
        for NavHead_tag in page_soup.find_all(name='div', attrs={'class': 'NavHead'}):
            if NavHead_tag.text == 'Influences':
                for a_tag in NavHead_tag.parent.find_all(name='a', title=True):
                    influences_title = a_tag.attrs['title']
                    influences_list.append(influences_title)
        if len(influences_list) == 0:
            for param_th in page_soup.find_all('th', scope='row'):
                if (param_th.text.strip()=='Influences'):
                    influences_td = param_th.find_next_sibling('td')
                    for influences_link in influences_td.find_all('a', title=True):
                        influences_title = influences_link['title']
                        influences_list.append(influences_title)
    except Exception as e:
        print(e)
    
    return influences_list

In [ ]:

bs = wikipedia.BeautifulSoup
import networkx as nx

butler_digraph = nx.DiGraph()
node_set = set()
root_node = 'Judith Butler'
def build_graph(influences_digraph, page_title, n=0):
    node_set.add(page_title)
    if len(node_set) < 100:
        influences_list = get_influences_list(page_title=page_title)
        not_in_node_set_list = []
        for influencer_title in influences_list:
            if influencer_title not in node_set:
                not_in_node_set_list.append(influencer_title)
                influences_digraph.add_node(influencer_title)
                influences_digraph.add_edge(influencer_title, page_title)
        for influencer_title in not_in_node_set_list:
            node_set.add(influencer_title)
            if (n < 100) and (nx.shortest_path_length(butler_digraph, influencer_title, root_node) < 4):
                influences_digraph = build_graph(influences_digraph=influences_digraph,
                                                 page_title=influencer_title, n=n+1)
    
    return influences_digraph

In [ ]:

import matplotlib.pyplot as plt
%matplotlib inline

# Draw the graph using the default spring layout
plt.figure(figsize=(18, 9))
butler_digraph = s.load_object('butler_digraph')
nx.draw_networkx(butler_digraph)

In [ ]:

# Draw the graph using the spring layout
plt.figure(figsize=(18, 9))
spring_pos = nx.spring_layout(butler_digraph)
nx.draw_networkx(butler_digraph, spring_pos)